src/main/java/org/threeten/extra/AmountFormats.java

```diff
@@ -287,6 +307,186 @@ private static boolean oppositeSigns(int a, int b) {
         return a < 0 ? (b >= 0) : (b < 0);
     }
 
+    // -------------------------------------------------------------------------
+    /**
+     * Parses a unit-based duration string to a duration value.
+     * <p>
+     * This is an implementation of Golang's {@link https://golang.org/pkg/time/#ParseDuration}:
+     * ParseDuration parses a duration string. A duration string is a possibly
+     * signed sequence of decimal numbers, each with optional fraction and a
+     * unit suffix, such as "300ms", "-1.5h" or "2h45m". Valid time units are
+     * "ns", "us" (or "µs"), "ms", "s", "m", "h".
+     * <p>
+     * Note, the value "0" is specially supported as {@code Duration.ZERO}, but
+     * unlike the Go implementation infinite durations are not supported.
+     *
+     * @param durationText the formatted unit-based duration string.
+     * @return the {@code Duration} value represented by the string, if possible.
+     */
+    public static Optional<Duration> parseUnitBasedDuration(CharSequence durationText) {
```

Comment:
The problem with returning optional is that no feedback is provided as to why the input is invalid. What was your thinking here? java.time.* generally throws exceptions in parsing.

Reply:
I was copying the `absl::optional` return behavior of the C++ implementation, but have switched to the more Java idiomatic exception behavior by throwing `java.time.DateTimeParseException` for invalid inputs.